File: app/src/main/java/com/ludoscity/findmybikes/fragments/FavoriteListFragment.java
package com.ludoscity.findmybikes.fragments;

import android.app.Activity;
import android.arch.lifecycle.Observer;
import android.arch.lifecycle.ViewModelProviders;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.helper.ItemTouchHelper;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.ludoscity.findmybikes.FavoriteRecyclerViewAdapter;
import com.ludoscity.findmybikes.ItemTouchHelperAdapter;
import com.ludoscity.findmybikes.R;
import com.ludoscity.findmybikes.datamodel.FavoriteEntityBase;
import com.ludoscity.findmybikes.datamodel.FavoriteEntityPlace;
import com.ludoscity.findmybikes.datamodel.FavoriteEntityStation;
import com.ludoscity.findmybikes.helpers.FavoriteRepository;
import com.ludoscity.findmybikes.utils.DividerItemDecoration;
import com.ludoscity.findmybikes.utils.ScrollingLinearLayoutManager;
import com.ludoscity.findmybikes.viewmodels.FavoriteListViewModel;
import com.ludoscity.findmybikes.viewmodels.NearbyActivityViewModel;

import java.util.List;

public class FavoriteListFragment extends Fragment implements
        FavoriteRecyclerViewAdapter.OnFavoriteListItemStartDragListener,//TODO: investigate making the sheet listening and forwarding
        FavoriteRecyclerViewAdapter.OnFavoriteListItemClickListener{

    public static final String FAVORITE_LIST_ITEM_CLICK_PATH = "station_list_item_click";
    public static final String FAVORITE_LIST_INACTIVE_ITEM_CLICK_PATH = "station_list_inactive_item_click";

    //private EditableMaterialSheetFab mFavoritesSheetFab;
    private NearbyActivityViewModel mNearbyActivityViewModel;
    private FavoriteListViewModel mFavoriteListViewModel;
    private ItemTouchHelper mFavoriteItemTouchHelper;
    FavoriteRecyclerViewAdapter mFavoriteRecyclerViewAdapter;

    private OnFavoriteListFragmentInteractionListener mListener;
    private boolean mIsFavoriteSheetEditInProgress = false;
    private boolean mIsFavoriteSheetItemNameEditInProgress = false;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        View inflatedView =  inflater.inflate(R.layout.fragment_favorite_list, container, false);

        //ItemTouchHelper.SimpleCallback simpleItemTouchCallback = new ItemTouchHelper.SimpleCallback(ItemTouchHelper.UP | ItemTouchHelper.DOWN,
        //        ItemTouchHelper.LEFT) {
        ItemTouchHelper.SimpleCallback simpleItemTouchCallback = new ItemTouchHelper.SimpleCallback( ItemTouchHelper.UP | ItemTouchHelper.DOWN
                , ItemTouchHelper.LEFT | ItemTouchHelper.RIGHT) {
            @Override
            public boolean onMove(RecyclerView recyclerView, RecyclerView.ViewHolder viewHolder, RecyclerView.ViewHolder target) {
                ((ItemTouchHelperAdapter)recyclerView.getAdapter()).onItemMove(viewHolder.getAdapterPosition(),
                        target.getAdapterPosition());
                return true;
            }

            @Override
            public void onSwiped(RecyclerView.ViewHolder viewHolder, int direction) {
                onFavoriteListItemDelete(((FavoriteRecyclerViewAdapter.FavoriteListItemViewHolder)viewHolder).getFavoriteId(), viewHolder.getAdapterPosition());
            }

            @Override
            public boolean isLongPressDragEnabled() {

                return mNearbyActivityViewModel.isFavoriteSheetEditInProgress().getValue();//mFavoriteRecyclerViewAdapter.getSheetEditing();

            }

            @Override
            public boolean isItemViewSwipeEnabled() {

                return !mIsFavoriteSheetEditInProgress && !mIsFavoriteSheetItemNameEditInProgress;
            }

            @Override
            public void onSelectedChanged(RecyclerView.ViewHolder _viewHolder, int _actionState){
                if (_actionState != ItemTouchHelper.ACTION_STATE_IDLE){
                    if (_viewHolder instanceof FavoriteRecyclerViewAdapter.FavoriteItemTouchHelperViewHolder){
                        FavoriteRecyclerViewAdapter.FavoriteItemTouchHelperViewHolder favoriteItemViewHolder =
                                (FavoriteRecyclerViewAdapter.FavoriteItemTouchHelperViewHolder) _viewHolder;
                        favoriteItemViewHolder.onItemSelected();
                    }
                }
                super.onSelectedChanged(_viewHolder, _actionState);
            }

            @Override
            public void clearView(RecyclerView _recyclerView, RecyclerView.ViewHolder _viewHolder){
                super.clearView(_recyclerView, _viewHolder);
                if (_viewHolder instanceof FavoriteRecyclerViewAdapter.FavoriteItemTouchHelperViewHolder){
                    FavoriteRecyclerViewAdapter.FavoriteItemTouchHelperViewHolder favoriteItemViewHolder =
                            (FavoriteRecyclerViewAdapter.FavoriteItemTouchHelperViewHolder) _viewHolder;
                    favoriteItemViewHolder.onItemClear();
                }
            }
        };

        mFavoriteItemTouchHelper = new ItemTouchHelper(simpleItemTouchCallback);

        RecyclerView favoriteRecyclerView = inflatedView.findViewById(R.id.favorite_list_recyclerview);

        favoriteRecyclerView.addItemDecoration(new DividerItemDecoration(getActivity(), DividerItemDecoration.VERTICAL_LIST));
        favoriteRecyclerView.setLayoutManager(new ScrollingLinearLayoutManager(getActivity(), LinearLayoutManager.VERTICAL, false, 300));

        mFavoriteListViewModel = ViewModelProviders.of(this).get(FavoriteListViewModel.class);
        mFavoriteListViewModel.getFavoriteEntityStationList().observe(this, new Observer<List<FavoriteEntityStation>>() {
            @Override
            public void onChanged(@Nullable List<FavoriteEntityStation> favoriteEntityStations) {
                //TODO: have more elaborate code, right now when a single item is changed, both this and the individual holders get notified
                //TODO: maybe add UI index comparison, or that will be taken care of by individual items observer pattern ?
                if (favoriteEntityStations.size() > mFavoriteRecyclerViewAdapter.getItemCount()) {
                    mFavoriteRecyclerViewAdapter.resetFavoriteList(favoriteEntityStations);
                }

                //TODO: investigate communicating the required state change through NearbyActivityViewModel
                //actually that should be done now by also observing the same model
                mListener.onFavoriteListChanged(favoriteEntityStations.size() == 0);
            }
        });

        mFavoriteRecyclerViewAdapter = new FavoriteRecyclerViewAdapter( this, this, getActivity().getApplicationContext(), mFavoriteListViewModel, this);

        favoriteRecyclerView.setAdapter(mFavoriteRecyclerViewAdapter);

        mFavoriteItemTouchHelper.attachToRecyclerView(favoriteRecyclerView);

        return inflatedView;
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

        mNearbyActivityViewModel = ViewModelProviders.of(getActivity()).get(NearbyActivityViewModel.class);

        mNearbyActivityViewModel.isFavoriteSheetEditInProgress().observe(this, new Observer<Boolean>() {
            @Override
            public void onChanged(@Nullable Boolean favSheetEditInProgress) {

                mIsFavoriteSheetEditInProgress = favSheetEditInProgress;
                mFavoriteRecyclerViewAdapter.setSheetEditing(mIsFavoriteSheetEditInProgress);
            }
        });

        mNearbyActivityViewModel.isFavoriteSheetItemNameEditInProgress().observe(this, new Observer<Boolean>() {
            @Override
            public void onChanged(@Nullable Boolean favItemNameEditInProgress) {

                mIsFavoriteSheetItemNameEditInProgress = favItemNameEditInProgress;
            }
        });
    }

    @Override
    public void onAttach(Activity activity) {
        super.onAttach(activity);
        try {
            mListener = (OnFavoriteListFragmentInteractionListener) activity;
        } catch (ClassCastException e) {
            throw new ClassCastException(activity.toString()
                    + " must implement OnFragmentInteractionListener");
        }
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mListener = null;
    }

    @Override
    public void onFavoriteListItemClick(String _stationId) {

        //TODO: act on NearbyActivityViewModel instead of a direct callback
        mListener.onFavoriteListItemClicked(_stationId);
    }

    @Override
    public void onFavoriteListItemNameEditDone(String _favoriteId, String _newName) {

        if (!_favoriteId.startsWith(FavoriteEntityPlace.PLACE_ID_PREFIX)) {

            FavoriteEntityBase fav = mFavoriteListViewModel.getFavoriteEntityForId(_favoriteId);
            fav.setCustomName(_newName);
            mFavoriteListViewModel.updateFavorite(fav);

            mListener.onFavoriteItemEditDone(_favoriteId);
        }
        else{
            FavoriteEntityBase favEntity = mFavoriteListViewModel.getFavoriteEntityForId(_favoriteId);
            CharSequence attr = favEntity.getAttributions();
            String attrString = "";
            if (attr != null)
                attrString = attr.toString();

            mFavoriteListViewModel.addFavorite(new FavoriteEntityPlace(favEntity.getId(), _newName, favEntity.getLocation(), attrString));
        }

        mNearbyActivityViewModel.showFavoriteSheetEditFab();
        mNearbyActivityViewModel.favoriteItemNameEditStop();
    }

    @Override
    public void onFavoriteListItemNameEditBegin() {
        mNearbyActivityViewModel.favoriteItemNameEditStart();

    }

    @Override
    public void onFavoristeListItemNameEditAbort() {

        //mFavoritesSheetFab.showEditFab();
        mNearbyActivityViewModel.favoriteItemNameEditStop();
        //mFavoriteItemEditInProgress = false;
    }

    @Override
    public void onFavoriteListItemDelete(String favoriteId, int adapterPosition) {
        mFavoriteRecyclerViewAdapter.removeFavorite(favoriteId, adapterPosition);

        mListener.onFavoriteItemDeleted(favoriteId, false);
        mFavoriteListViewModel.removeFavorite(favoriteId);
    }

    @Override
    public void onFavoriteListItemStartDrag(RecyclerView.ViewHolder _viewHolder){
        mFavoriteItemTouchHelper.startDrag(_viewHolder);
    }

    public interface OnFavoriteListFragmentInteractionListener {

        void onFavoriteItemEditDone(String fsvoriteId);
        void onFavoriteItemDeleted(String favoriteId, boolean showUndo);
        void onFavoriteListChanged(boolean noFavorite);

        void onFavoriteListItemClicked(String favoriteId);
    }

}

File: app/src/main/java/com/ludoscity/findmybikes/ui/sheet/FavoriteListFragment.kt
package com.ludoscity.findmybikes.ui.sheet

import android.arch.lifecycle.Observer
import android.arch.lifecycle.ViewModelProviders
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.support.v7.widget.helper.ItemTouchHelper
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import com.ludoscity.findmybikes.R
import com.ludoscity.findmybikes.ui.main.FindMyBikesActivityViewModel
import com.ludoscity.findmybikes.utils.InjectorUtils

class FavoriteListFragment : Fragment(), FavoriteRecyclerViewAdapter.OnFavoriteListItemStartDragListener, //TODO: investigate making the sheet listening and forwarding
        FavoriteRecyclerViewAdapter.OnFavoriteListItemClickListener, EditableMaterialSheetFab.OnFavoriteSheetEventListener {

    //private EditableMaterialSheetFab mFavoritesSheetFab;
    private var findMyBikesActivityViewModel: FindMyBikesActivityViewModel? = null
    private var favoriteSheetListViewModel: FavoriteSheetListViewModel? = null
    private var favoriteItemTouchHelper: ItemTouchHelper? = null
    private lateinit var favoriteRecyclerViewAdapter: FavoriteRecyclerViewAdapter
    private lateinit var favoriteRecyclerView: RecyclerView

    private var listener: OnFavoriteListFragmentInteractionListener? = null
    private var isFavoriteSheetEditInProgress = false
    private var isFavoriteSheetItemNameEditInProgress = false

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        // Inflate the layout for this fragment
        val inflatedView = inflater.inflate(R.layout.fragment_favorite_list, container, false)

        favoriteRecyclerView = inflatedView.findViewById(R.id.favorite_list_recyclerview)

        return inflatedView
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)

        findMyBikesActivityViewModel = ViewModelProviders.of(activity!!).get(FindMyBikesActivityViewModel::class.java)

        findMyBikesActivityViewModel!!.isFavoriteSheetItemNameEditInProgress.observe(
                this,
                Observer { favItemNameEditInProgress ->
                    isFavoriteSheetItemNameEditInProgress = favItemNameEditInProgress!!
                })

        val facto = InjectorUtils.provideFavoriteSheetListFragmentViewModelFactory(activity!!.application, findMyBikesActivityViewModel!!.isFavoriteSheetEditInProgress)

        favoriteSheetListViewModel = ViewModelProviders.of(this, facto).get(FavoriteSheetListViewModel::class.java)

        favoriteSheetListViewModel!!.sheetItemDataList.observe(this, Observer { favoriteSheetItemData -> favoriteRecyclerViewAdapter.loadItems(favoriteSheetItemData!!) })

        favoriteRecyclerViewAdapter = FavoriteRecyclerViewAdapter(this, this, activity!!.applicationContext)

        favoriteRecyclerView.addItemDecoration(DividerItemDecoration(activity!!, DividerItemDecoration.VERTICAL_LIST))
        favoriteRecyclerView.layoutManager = ScrollingLinearLayoutManager(activity!!, LinearLayoutManager.VERTICAL, false, 300)

        //ItemTouchHelper.SimpleCallback simpleItemTouchCallback = new ItemTouchHelper.SimpleCallback(ItemTouchHelper.UP | ItemTouchHelper.DOWN,
        //        ItemTouchHelper.LEFT) {
        val simpleItemTouchCallback = object : ItemTouchHelper.SimpleCallback(ItemTouchHelper.UP or ItemTouchHelper.DOWN, ItemTouchHelper.LEFT or ItemTouchHelper.RIGHT) {
            override fun onMove(recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder, target: RecyclerView.ViewHolder): Boolean {
                (recyclerView.adapter as ItemTouchHelperAdapter).onItemMove(viewHolder.adapterPosition,
                        target.adapterPosition)
                return true
            }

            override fun onSwiped(viewHolder: RecyclerView.ViewHolder, direction: Int) {
                onFavoriteListItemDelete((viewHolder as FavoriteRecyclerViewAdapter.FavoriteListItemViewHolder).favoriteId!!, viewHolder.getAdapterPosition())
            }

            override fun isLongPressDragEnabled(): Boolean {
                return findMyBikesActivityViewModel!!.isFavoriteSheetEditInProgress.value!!
            }

            override fun isItemViewSwipeEnabled(): Boolean {
                return !isFavoriteSheetEditInProgress && !isFavoriteSheetItemNameEditInProgress
            }

            override fun onSelectedChanged(viewHolder: RecyclerView.ViewHolder?, actionState: Int) {
                if (actionState != ItemTouchHelper.ACTION_STATE_IDLE) {
                    if (viewHolder is FavoriteRecyclerViewAdapter.FavoriteItemTouchHelperViewHolder) {
                        val favoriteItemViewHolder = viewHolder as FavoriteRecyclerViewAdapter.FavoriteItemTouchHelperViewHolder?
                        favoriteItemViewHolder!!.onItemSelected()
                    }
                }
                super.onSelectedChanged(viewHolder, actionState)
            }

            override fun clearView(recyclerView: RecyclerView?, viewHolder: RecyclerView.ViewHolder) {
                super.clearView(recyclerView, viewHolder)
                if (viewHolder is FavoriteRecyclerViewAdapter.FavoriteItemTouchHelperViewHolder) {
                    val favoriteItemViewHolder = viewHolder as FavoriteRecyclerViewAdapter.FavoriteItemTouchHelperViewHolder
                    favoriteItemViewHolder.onItemClear()
                }
            }
        }

        favoriteItemTouchHelper = ItemTouchHelper(simpleItemTouchCallback)

        favoriteRecyclerView.adapter = favoriteRecyclerViewAdapter

        favoriteItemTouchHelper!!.attachToRecyclerView(favoriteRecyclerView)

        try {
            listener = activity as OnFavoriteListFragmentInteractionListener?
        } catch (e: Exception) {
            throw ClassCastException(activity!!.toString() + " must implement OnFragmentInteractionListener")
        }

    }

    override fun onDetach() {
        super.onDetach()
        listener = null
    }

    override fun onFavoriteListItemClick(favoriteId: String) {
        findMyBikesActivityViewModel!!.setLastClickedFavoriteListItemFavoriteId(favoriteId)
    }

    override fun onFavoriteListItemNameEditDone(favoriteId: String, newName: String) {

        if (!newName.isEmpty()) {
            favoriteSheetListViewModel!!.updateFavoriteCustomNameByFavoriteId(favoriteId, newName)
            findMyBikesActivityViewModel!!.refreshStationAAndB()
        }

        findMyBikesActivityViewModel!!.showFavoriteSheetEditFab()
        findMyBikesActivityViewModel!!.favoriteItemNameEditStop()
    }

    override fun onFavoriteListItemNameEditBegin() {
        findMyBikesActivityViewModel!!.favoriteItemNameEditStart()
    }

    override fun onFavoristeListItemNameEditAbort() {
        findMyBikesActivityViewModel!!.favoriteItemNameEditStop()
    }

    override fun onFavoriteListItemDelete(favoriteId: String, adapterPosition: Int) {
        favoriteSheetListViewModel!!.removeFavorite(favoriteId)
    }

    override fun onFavoriteListItemStartDrag(viewHolder: RecyclerView.ViewHolder) {
        favoriteItemTouchHelper!!.startDrag(viewHolder)
    }

    override fun onFavoriteSheetEditDone() {

        for ((i, fav) in favoriteRecyclerViewAdapter.items.withIndex()) {
            favoriteSheetListViewModel!!.updateFavoriteUiIndexByFavoriteId(fav.favoriteId, i)
        }
        findMyBikesActivityViewModel!!.favoriteSheetEditDone()
    }

    override fun onFavoriteSheetEditCancel() {
    }

    interface OnFavoriteListFragmentInteractionListener {

        fun onFavoriteItemDeleted(favoriteId: String, showUndo: Boolean)
        fun onFavoriteListChanged(noFavorite: Boolean)

        fun onFavoriteListItemClicked(favoriteId: String)
    }

    companion object {

        val FAVORITE_LIST_ITEM_CLICK_PATH = "station_list_item_click"
        val FAVORITE_LIST_INACTIVE_ITEM_CLICK_PATH = "station_list_inactive_item_click"
    }

}
